python/servo/mutation/mutator.py

```diff
@@ -0,0 +1,203 @@
+# Copyright 2013 The Servo Project Developers. See the COPYRIGHT
+# file at the top-level directory of this distribution.
+#
+# Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+# http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+# <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+# option. This file may not be copied, modified, or distributed
+# except according to those terms.
+
+import fileinput
+import re
+import random
+
+regex = {
+    'arithmetic_plus': r'(?<=[^\"]\s)\+(?=\s[^A-Z\'?\":\{]+)',
+    'arithmetic_minus': r'(?<=\s)\-(?=\s.+)',
+    'plus_equals': r'(?<=\s)\+(?=\=)',
+    'minus_equals': r'(?<=\s)\-(?=\=)',
+    'push_statement': r'.*?push\(.*?\).*?;',
+    'remove_statement': r'.*?remove\(.*?\).*?;',
+    'append_statement': r'.*?append\(.*?\).*?;',
+    'atomic_string': r"(?<=\").+(?=\")",
+    'if_condition': r'(?<=if\s)(.*)(?=\s\{)',
+    'logical_and': r'\s&&\s',
+    'less_than_equals': r'(?<=\s\<)\=\s',
+    'greater_than_equals': r'(?<=\s\>)\=\s',
+    'if_block': r'[^a-z]\sif(.+)\{',
+    'else_block': r'\selse(.+)\{'
+}
+
+
+def is_comment(line):
+    return re.search(r'\/\/.*', line)
+
+
+def init_variables(if_blocks):
+    random_index = random.randint(0, len(if_blocks) - 1)
+    start_counter = 0
+    end_counter = 0
+    lines_to_delete = []
+    line_to_mutate = if_blocks[random_index]
+    return random_index, start_counter, end_counter, lines_to_delete, line_to_mutate
+
+
+def deleteStatements(file_name, line_numbers):
+    for line in fileinput.input(file_name, inplace=True):
+        if fileinput.lineno() not in line_numbers:
+            print line.rstrip()
+
+
+class Strategy:
+    def __init__(self):
+        self._strategy_name = ""
+        self._replace_strategy = {}
+
+    def mutate(self, file_name):
+        line_numbers = []
+        for line in fileinput.input(file_name):
+            if not is_comment(line) and re.search(self._replace_strategy['regex'], line):
+                line_numbers.append(fileinput.lineno())
+        if len(line_numbers) == 0:
+            return -1
+        else:
+            mutation_line_number = line_numbers[random.randint(0, len(line_numbers) - 1)]
+            for line in fileinput.input(file_name, inplace=True):
+                if fileinput.lineno() == mutation_line_number:
+                    if self._replace_strategy['replaceString']:
+                        line = re.sub(self._replace_strategy['regex'], self._replace_strategy['replaceString'], line)
+                    else:
+                        if self._strategy_name == "duplicate":
+                            replacement = line + line
```

Comment:
Hard-wiring in duplication seems like it could be done more cleanly., perhaps by having the replacement text referring to the match?